lib/semantics/statements.cc

```diff
@@ -0,0 +1,97 @@
+#include "statements.h"
+#include "semantics.h"
+#include "../common/indirection.h"
+#include "../common/Fortran.h"
+#include "../evaluate/expression.h"
+#include "../parser/parse-tree.h"
+#include <optional>
+#include <variant>
+
+namespace Fortran::semantics {
+
+class StatementContext {
+public:
+  explicit StatementContext(
+      SemanticsContext &c, parser::CharBlock at = parser::CharBlock{})
+    : context_{c}, messages_{at, &c.messages()} {}
+
+  bool operator==(const StatementContext &x) const { return this == &x; }
+  bool operator!=(const StatementContext &x) const { return this != &x; }
+
+  void Analyze(const parser::StopStmt &stmt) {
+    const auto &sc{std::get<std::optional<parser::StopCode>>(stmt.t)};
+    const auto &sle{std::get<std::optional<parser::ScalarLogicalExpr>>(stmt.t)};
+
+    if (sc) {
+      const parser::CharBlock &source{sc.value().v.thing.source};
+      const auto &type{sc.value().v.thing.typedExpr->v.GetType()};
+
+      if (!type) {
+        messages_.Say(source,
+          "type information not ready for STOP code"_err_en_US);
```

Comment:
This will appear like a user error, but it's not. It's something that's not implemented yet. A `CHECK` is probably more appropriate.

Reply:
I assume I need to give up on that 'if' condition and allow it to fail (until missing bit is implemented). How `CHECK` should be used to handle situation like that?

Reply:
I was thinking something like this:
```
CHECK(type && "no type for stop code");
```

Reply:
I'd shift the burden on tools.h functions and see if you like it better.